subsys/net/lib/lwm2m/lwm2m_rw_cbor.c

```diff
@@ -0,0 +1,559 @@
+	CBOR_TAG_TIME_TSTR =    0, /* text string	Standard date/time string */
+	CBOR_TAG_TIME_NUM =     1, /* integer or float	Epoch-based date/time */
+	CBOR_TAG_UBIGNUM_BSTR = 2, /* byte string	Unsigned bignum */
+	CBOR_TAG_BIGNUM_BSTR =  3, /* byte string	Negative bignum */
+	CBOR_TAG_DECFRAC_ARR =  4, /* array		Decimal fraction */
+	CBOR_TAG_BIGFLOAT_ARR = 5, /* array		Bigfloat */
+	CBOR_TAG_2BASE64URL =  21, /* (any)		Expected conversion to base64url encoding */
+	CBOR_TAG_2BASE64 =     22, /* (any)		Expected conversion to base64 encoding */
+	CBOR_TAG_2BASE16 =     23, /* (any)		Expected conversion to base16 encoding */
+	CBOR_TAG_BSTR =        24, /* byte string	Encoded CBOR data item */
+	CBOR_TAG_URI_TSTR =    32, /* text string	URI */
+	CBOR_TAG_BASE64URL_TSTR = 33, /* text string	base64url */
+	CBOR_TAG_BASE64_TSTR =    34, /* text string	base64 */
+	CBOR_TAG_MIME_TSTR =      36, /* text string	MIME message */
+	CBOR_TAG_CBOR =        55799, /* (any)	Self-described CBOR */
+};
+
+static size_t put_time(struct lwm2m_output_context *out, struct lwm2m_obj_path *path, int64_t value)
+{
+	/* CBOR time output format is unspecified but SenML CBOR uses string format.
+	 * Let's stick into the same format with plain CBOR
+	 */
+	struct tm dt;
+	char time_str[sizeof("1970-01-01T00:00:00-00:00")] = { 0 };
+	int len;
+	int str_sz;
+	int tag_sz;
+
+	if (gmtime_r((time_t *)&value, &dt) != &dt) {
+		LOG_ERR("unable to convert from secs since Epoch to a date/time construct");
+		return -EINVAL;
+	}
+
+	/* Time construct to a string. Time in UTC, offset to local time not known */
+	len = snprintk(time_str, sizeof(time_str),
+			"%04d-%02d-%02dT%02d:%02d:%02d-00:00",
+			dt.tm_year+1900,
+			dt.tm_mon+1,
+			dt.tm_mday,
+			dt.tm_hour,
+			dt.tm_min,
+			dt.tm_sec);
+
+	if (len < 0 || len > sizeof(time_str) - 1) {
+		LOG_ERR("unable to form a date/time string");
+		return -EINVAL;
+	}
+
+	cbor_state_t states[1];
+
+	new_state(states, 1, CPKT_BUF_W_REGION(out->out_cpkt), 1);
+
+	/* TODO: Are tags required? V1.1 leaves this unspecified but some servers require tags */
+	bool ret = tag_encode(states, CBOR_TAG_TIME_TSTR);
+
+	if (!ret) {
+		LOG_ERR("unable to encode date/time string tag");
+		return -ENOMEM;
+	}
+
+	tag_sz = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+
+	out->out_cpkt->offset += tag_sz;
+
+	ret = tstrx_encode(states,
+			   &(cbor_string_type_t){ .value = time_str, .len = strlen(time_str) });
+	if (!ret) {
+		LOG_ERR("unable to encode date/time string");
+		return -ENOMEM;
+	}
+
+	str_sz = CPKT_CBOR_W_SZ(states[0].payload, out->out_cpkt);
+
+	out->out_cpkt->offset += str_sz;
+
+	return tag_sz + str_sz;
+}
+
+static size_t put_s64(struct lwm2m_output_context *out, struct lwm2m_obj_path *path, int64_t value)
+{
+	cbor_state_t states[1];
+	int payload_len;
+
+	/* TODO: remove this check once the CBOR library supports 64-bit values */
+	if (value != (value & UINT32_MAX)) {
```

Comment:
What is this comparison trying to do?
This looks like a check that our value can fit into `uint32_t`, but it used binary AND operation for that.
However, as signed negative values are two's complement, `-1` is same as `0xFFFFFFFFFFFFFFFF` when in `int64_t` so it looks like it is not fitting. but equally `int32_t (-1) = 0xFFFFFFFF` which should fit.

So is it that we only support unsigned 32 bit? or do we have signed 32 bits as well? If we do, we should have
```
if (value > INT32_MAX || value < INT32_MIN) {
    LOG_WRN(...);
    return -EINVAL;
}
/* Encode as signed 32 bit, until the library supports 64-bit values */
return put_s32(out, path, (int32_t) value);
```
We have `intx32_encode()` right?

Reply:
For now I only intend to use this function for handling uint32_t values. This will change once the zbor's 64-bit support is available. And yes, we do have `intx32_encode()`